x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -275,221 +284,109 @@ public void checkCompatibility(MappedFieldType otherFT, List<String> conflicts)
-                    }
-                    i += length;
-                }
-                if (sb.length() > 0) {
-                    precedingGapSizeList.add(lastGap);
-                    fragmentList.add(sb.toString());
-                    lastGap = 0;
-                }
-                fragments = fragmentList.toArray(new String[0]);
-                precedingGapSizes = precedingGapSizeList.toArray(new Integer[0]);
-
-            }
-
-            public boolean needsVerification() {
-                // Return true if term queries are not enough evidence
-                if (fragments.length == 1 && wildcardCharCount == 0) {
-                    // The one case where we don't need verification is when
-                    // we have a single fragment and no ? characters
-                    return false;
-                }
-                return true;
-            }
-
-            // Returns number of positions for last gap (Integer.MAX means unlimited gap)
-            public int getPrecedingGapSize(int fragmentNum) {
-                return precedingGapSizes[fragmentNum];
-            }
-
-            public boolean isMatchAll() {
-                return fragments.length == 0 && wildcardStringCount >0 && wildcardCharCount ==0;
-            }
-
-            @Override
-            public int hashCode() {
-                return pattern.hashCode();
-            }
-
-            @Override
-            public boolean equals(Object obj) {
-                PatternStructure other = (PatternStructure) obj;
-                return pattern.equals(other.pattern);
-            }
-
-
-        }
-
-
         @Override
         public Query wildcardQuery(String wildcardPattern, RewriteMethod method, QueryShardContext context) {
+            
             if (normalizer != null) {
                 wildcardPattern = StringFieldType.normalizeWildcardPattern(name(), wildcardPattern, normalizer);
             }
-            PatternStructure patternStructure = new PatternStructure(wildcardPattern);
-            ArrayList<String> tokens = new ArrayList<>();
-
-            for (int i = 0; i < patternStructure.fragments.length; i++) {
-                String fragment = patternStructure.fragments[i];
-                int fLength = fragment.length();
-                if (fLength == 0) {
-                    continue;
-                }
+            Automaton dvAutomaton = WildcardQuery.toAutomaton(new Term(name(), wildcardPattern));
+            
+            //Unlike the doc values, ngram index has extra string-start and end characters. 
+            String ngramWildcardPattern = TOKEN_START_OR_END_CHAR + wildcardPattern 
+                    + TOKEN_START_OR_END_CHAR + TOKEN_START_OR_END_CHAR;            
+            Automaton ngramAutomaton = WildcardQuery.toAutomaton(new Term(name(), ngramWildcardPattern));
+            return automatonToQuery(wildcardPattern, wildcardPattern, ngramAutomaton, dvAutomaton);
```

Comment:
Do you think it is worth catching the `TooComplex` exception every time you call `automatonToQuery`? Maybe it should be a checked exception (gasp!) and we always catch it and rewrite it to `IllegalArgumentExcpetion` with helpful words.